rustls/src/crypto/hpke.rs

```diff
@@ -0,0 +1,131 @@
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+use core::fmt::Debug;
+
+use crate::msgs::enums::{HpkeAead, HpkeKdf, HpkeKem};
+use crate::Error;
+
+/// A provider for [RFC 9180] Hybrid Public Key Encryption (HPKE).
+///
+/// At a minimum each provider must support base mode and the [HPKE ciphersuite profile] required for
+/// encrypted client hello (ECH):
+///  * KEM: DHKEM(X25519, HKDF-SHA256)
+///  * KDF: HKDF-SHA256
+///  * AEAD: AES-128-GCM
+///
+/// [RFC 9180]: <https://www.rfc-editor.org/rfc/rfc9180.html>
+/// [HPKE ciphersuite profile]: <https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-17#section-9>
+pub trait HpkeProvider: Debug + Send + Sync + 'static {
+    /// Create an HPKE instance for the given mode, KEM, KDF and AEAD.
+    ///
+    /// May return an error if the mode, KEM, KDF or AEAD are unsupported by the HPKE provider.
+    fn setup(
+        &self,
+        mode: Mode,
+        kem: HpkeKem,
+        kdf: HpkeKdf,
+        aead: HpkeAead,
+    ) -> Result<Box<dyn Hpke>, Error>;
+
+    /// Does the provider support the given mode?
+    fn supports_mode(&self, mode: Mode) -> bool;
```

Comment:
What is the utility of these `supports_*()` methods? It feels like either they should be more like getters yielding a slice of supported elements, or, if there's not much we can do if we don't support a particular element, we should leverage the fallibility of the `setup()` method.

Reply:
> What is the utility of these supports_*() methods? 

For the ECH use-case we'll be working with config structures that end up describing [`HpkeKeyConfig`](https://github.com/rustls/rustls/blob/ccb79947a4811412ee7dcddcd0f51ea56bccf101/rustls/src/msgs/handshake.rs#L2279-L2284) instances, that each contain a `Vec<HpkeSymmetricCipherSuite>`. Each [`HpkeSymmetricCipherSuite`](https://github.com/rustls/rustls/blob/ccb79947a4811412ee7dcddcd0f51ea56bccf101/rustls/src/msgs/handshake.rs#L2255-L2258) has KEM, KDF, and AEAD IDs. We'll need a way to match what we support vs what the ECH config advertises to arrive at a compatible suite to use.

>  It feels like either they should be more like getters yielding a slice of supported elements

I had something like this initially, and this might be a place where I really am over-fitting hpke-rs since it doesn't support getting a list. I thought about hardcoding a list, but that feels too fragile.

I think we'd end up doing `contains` testing on a per-element basis against ECH config elements in either case, so making the interface express that "Do you support X" idea vs getting a list of all supported and iterating seemed OK. WDYT?